### ISSUE TITLE
::critcl::load does not return error when compile fails

### ISSUE THREAD

**Issue by tpoindex** [andreas-kupries/critcl#53]
I am using `::critcl::ccode` and `::critcl::ccommand` to define a Tcl command, and invoke `::critcl::load` to compile and load.  I have also defined `::critcl::cflags -Werror=return-type` to have gcc raise an error when a return type does not match the function definition.  `::critcl::load` does not raise a Tcl error when the compilation fails.

Here is a sample to demostrate the behavior:

```
package require critcl 3.1.13
set name bar
set helper_function {Tcl_WideInt TSP_UserDirect_foo() {}}
set tcl_cmd_interface {TSP_UserDirect_foo(); return TCL_OK;}
set rc [ catch {
    ::critcl::tcl 8.5
    ::critcl::cflags -Werror=return-type
    ::critcl::ccode  $helper_function
    ::critcl::ccommand ::$name {clientData interp objc objv} $tcl_cmd_interface
    ::critcl::load
} result ]
puts "rc = $rc"
```

Gcc will cause the compile to fail with the `-Werror=return-type` flag, and the critcl log for the compile shows the compile failure:

```
.critcl/linux-ix86/v3113_6f060491c56fdbd9f4088125bc1a3a79.c: In function ‘TSP_UserDirect_foo’:
.critcl/linux-ix86/v3113_6f060491c56fdbd9f4088125bc1a3a79.c:10:1: error: control reaches end of non-void function [-Werror=return-type]
 Tcl_WideInt TSP_UserDirect_foo() {}
 ^
cc1: some warnings being treated as errors
ERROR while compiling code in :
child process exited abnormally

critcl build failed ()
```

In the above Tcl code, the 'catch' block returns 0 as rc, and a Tcl error is not raised.  

**Expected behavior:**
::critcl::load should return 1, and raise a Tcl error.

**Comment by andreas-kupries** [comment]
critcl::load's result is the negation of critcl::cbuild.
Will have to dig deeper into that one.
As a general theory, the gcc error is somehow not detected as a build failure.
The main thing to find out is why.
The example you provided should help with that, thank you.

**Comment by andreas-kupries** [comment]
Investigation complete.
I have to reject this as a bug.
The build failure is actually detected.
And the result as seen is correct.

Because "critcl::load" behaves as it is [documented](http://andreas-kupries.github.io/critcl/doc/files/critcl_pkg.html#42) to.

Its return value is a boolean. It returns *true* on successful build, and *false* on failure.
It does _not_ throw an error on build failure, as your report apparently assumes.

At best this ticket is a request to change the public API of "critcl::load"
That would be a major version version.

It would be possible to avoid that by introducing a variant of load, i.e. under a different name ("load*" ?), which has the requested API, i.e. throw a Tcl error on failure, and an empty result otherwise.

**Comment by tpoindex** [comment]
Thanks for investigating,  I obviously didn't understand the docs for ::critcl::load.  Having critcl::load return '0' for failure is fine for me, no need for an api change.

I still have a related issue, after a compile failure, this dict returned by critcl::cresults has an empty 'log' key.   I'm not clear what the man page for critcl::cresults means - is 'log' only populated after a failure, and only when the build is triggered by ::critcl::buildforpackage ?  

I'd like to have the log contain just the compiler messages.  Having re-defined ::critcl::print as per docs to capture logging from critcl and parsing the log messages for my project, I'd like to have some log that is just the compiler log and not the information that critcl prints, e.g.,   timestamp & filename, compiler used, compiler flags, the compiler output, a Tcl error, and a few interspersed whitespace lines.

Test case:

```
package require critcl
critcl::cproc fudge  {char* s} char*  {return booger ;}
set result [critcl::load]
set result_dict [critcl::cresults]
puts "result: $result"
puts "   log: [dict get $result_dict log]"
```

whereas the ::critcl::print command prints:
```
Tue May 05 20:09:58 MDT 2015 - /home/tpoindex/crit-load-test.tcl

gcc
	-c
	-fPIC
	-DUSE_THREAD_ALLOC=1
	-D_REENTRANT=1
	-D_THREAD_SAFE=1
	-DHAVE_PTHREAD_ATTR_SETSTACKSIZE=1
	-DHAVE_READDIR_R=1
	-DTCL_THREADS=1
	-DUSE_TCL_STUBS
	-I/home/tpoindex/src/tcl/critcl-3.1.13-test/lib/critcl/critcl_c/tcl8.4
	-I/home/tpoindex/.critcl/linux-ix86
	-o
	/home/tpoindex/.critcl/linux-ix86/v3113_9a72bf1aaa8aec9265bd35f7d545d587_pic.o
	/home/tpoindex/.critcl/linux-ix86/v3113_9a72bf1aaa8aec9265bd35f7d545d587.c
	-O2
	-DNDEBUG

crit-load-test.tcl: In function ‘c__fudge0’:
crit-load-test.tcl:3:8: error: ‘booger’ undeclared (first use in this function)
crit-load-test.tcl:3:8: note: each undeclared identifier is reported only once for each function it appears in
ERROR while compiling code in /home/tpoindex/crit-load-test.tcl:
child process exited abnormally

critcl build failed (/home/tpoindex/crit-load-test.tcl)

```

What I really want is just the three lines of compiler errors.  Parsing these out is obviously compiler dependent.

**Comment by andreas-kupries** [comment]
> Thanks for investigating, I obviously didn't understand the docs for ::critcl::load. 

Having re-read the docs I believe that they could be better. Will clarify them.

Regarding your related issue, I have copied that into a new ticket, see #54.

**Comment by andreas-kupries** [comment]
With the docs updated I can now close this one.
